train.py

```diff
@@ -0,0 +1,230 @@
+    discriminator.train()
+
+    def random_labels(batch_size, start, end, device):
+        return (end - start) * torch.rand(batch_size, device=device) + start
+
+    fig_list = []
+
+    progress_bar = tqdm(enumerate(train_loader), total=len(train_loader))
+    for batch_idx, batch in progress_bar:
+        real_data = batch[0].to(cfg.system.device)
+        batch_size = real_data.size(0)
+
+        real_labels = random_labels(batch_size, 0.8, 1, cfg.system.device)
+        fake_labels = random_labels(batch_size, 0.0, 0.1, cfg.system.device)
+
+        # train discriminator
+        discriminator.zero_grad()
+        label = real_labels
+        real_data = real_data.view(real_data.shape[0], 1, real_data.shape[1]).to(cfg.system.device)
+        disc_real_output = discriminator(real_data).view(-1)
+        discriminator_error_real = criterion(disc_real_output, label)
+        discriminator_error_real.backward()
+        D_x = disc_real_output.mean().item()  # Mean discriminator output for real data
+
+        # train generator
+        noise = torch.randn(batch_size, cfg.generator.nz, cfg.generator.output_channels, device=cfg.system.device)
+        fake_data = generator(noise)
+        label = fake_labels
+        disc_fake_output = discriminator(fake_data.detach()).view(-1)
+
+        discriminator_error_fake = criterion(disc_fake_output, label)
+        discriminator_error_fake.backward()
+        D_G_z1 = disc_fake_output.mean().item()  # Discriminator's average output when evaluating the fake data
+        discriminator_error = discriminator_error_real + discriminator_error_fake
+        disc_optimizer.step()
+        generator.zero_grad()
+
+        label = real_labels
+        disc_output_after_update = discriminator(fake_data).view(-1)
+        generator_error = criterion(disc_output_after_update, label)
+        generator_error.backward()
+        D_G_z2 = disc_output_after_update.mean().item()  # Discriminator's output after updating the generator, fake data
+        gen_optimizer.step()
+
+        # log to wandb
+        wandb.log(
+            {
+                "generator_error": generator_error.item(),
+                "discriminator_error": discriminator_error.item(),
+                "D_x": D_x,
+                "D_G_z1": D_G_z1,
+                "D_G_z2": D_G_z2,
+            }
+        )
+        if batch_idx % cfg.train.log_interval == 0:
+            fig = visualize_training(generator, fixed_noise, epoch, batch_idx, cfg.logger.chart_path)
+            fig_list.append(fig)
+
+    checkpoint = {
+        "epoch": epoch,
+        "generator_state_dict": generator.state_dict(),
+        "discriminator_state_dict": discriminator.state_dict(),
+        "gen_optimizer_state_dict": gen_optimizer.state_dict(),
+        "disc_optimizer_state_dict": disc_optimizer.state_dict(),
+        "config": cfg,
+        "fixed_noise": fixed_noise,
+    }
+    torch.save(checkpoint, f"{cfg.logger.checkpoint_path}GAN_epoch_{epoch}.pth")
```

Comment:
```suggestion
    torch.save(checkpoint, f"{cfg.logger.checkpoint_path}/{cfg.run_name}_{epoch}.pth")
```

You'll need to add something like this to the config:

```yaml
run_name: ecg-gan-${now:%Y-%m-%d-%H-%M}
```

Reply:
https://stackoverflow.com/questions/59095824/what-is-the-difference-between-pt-pth-and-pwf-extentions-in-pytorch
According to this, also changed the extenstion to a one that isn't used by Python path configuration files.